File: python/pytoon/decode/validation.py
from __future__ import annotations

from pytoon.constants import COLON, LIST_ITEM_PREFIX
from pytoon.types import ArrayHeaderInfo, BlankLineInfo, Depth, ResolvedDecodeOptions
from pytoon.decode.scanner import LineCursor


class RangeError(ValueError):
    """Mirror the JS RangeError semantics."""
    pass


def assert_expected_count(actual: int, expected: int, item_type: str, options: ResolvedDecodeOptions) -> None:
    if options.strict and actual != expected:
        raise RangeError(f"Expected {expected} {item_type}, but got {actual}")


def validate_no_extra_list_items(cursor: LineCursor, item_depth: Depth, expected_count: int) -> None:
    next_line = cursor.peek()
    if next_line and next_line.depth == item_depth and next_line.content.startswith(LIST_ITEM_PREFIX):
        raise RangeError(f"Expected {expected_count} list array items, but found more")


def validate_no_extra_tabular_rows(cursor: LineCursor, row_depth: Depth, header: ArrayHeaderInfo) -> None:
    next_line = cursor.peek()
    if next_line and next_line.depth == row_depth and not next_line.content.startswith(LIST_ITEM_PREFIX):
        if _is_data_row(next_line.content, header.delimiter):
            raise RangeError(f"Expected {header.length} tabular rows, but found more")


def validate_no_blank_lines_in_range(
    start_line: int,
    end_line: int,
    blank_lines: list[BlankLineInfo],
    strict: bool,
    context: str,
) -> None:
    if not strict:
        return
    for blank in blank_lines:
        if start_line < blank.line_number < end_line:
            raise SyntaxError(
                f"Line {blank.line_number}: Blank lines inside {context} are not allowed in strict mode",
            )


def _is_data_row(content: str, delimiter: str) -> bool:
    colon_pos = content.find(COLON)
    delimiter_pos = content.find(delimiter)
    if colon_pos == -1:
        return True
    if delimiter_pos != -1 and delimiter_pos < colon_pos:
        return True
    return False

File: packages/toon/src/decode/validation.ts
import type { ArrayHeaderInfo, BlankLineInfo, Delimiter, Depth, ParsedLine } from '../types.ts'
import { COLON, LIST_ITEM_PREFIX } from '../constants.ts'
import { ToonDecodeError } from './errors.ts'

// #region Count and structure validation

/**
 * Asserts that the actual count matches the expected count in strict mode.
 */
export function assertExpectedCount(
  actual: number,
  expected: number,
  itemType: string,
  options: { strict: boolean },
  line: ParsedLine,
): void {
  if (options.strict && actual !== expected) {
    throw new ToonDecodeError(
      `Expected ${expected} ${itemType}, but got ${actual}`,
      { line: line.lineNumber, source: line.raw },
    )
  }
}

/**
 * Validates that there are no extra list items beyond the expected count.
 */
export function validateNoExtraListItems(
  nextLine: ParsedLine | undefined,
  itemDepth: Depth,
  expectedCount: number,
): void {
  if (nextLine?.depth === itemDepth && nextLine.content.startsWith(LIST_ITEM_PREFIX)) {
    throw new ToonDecodeError(
      `Expected ${expectedCount} list array items, but found more`,
      { line: nextLine.lineNumber, source: nextLine.raw },
    )
  }
}

/**
 * Validates that there are no extra tabular rows beyond the expected count.
 */
export function validateNoExtraTabularRows(
  nextLine: ParsedLine | undefined,
  rowDepth: Depth,
  header: ArrayHeaderInfo,
): void {
  if (
    nextLine?.depth === rowDepth
    && !nextLine.content.startsWith(LIST_ITEM_PREFIX)
    && isDataRow(nextLine.content, header.delimiter)
  ) {
    throw new ToonDecodeError(
      `Expected ${header.length} tabular rows, but found more`,
      { line: nextLine.lineNumber, source: nextLine.raw },
    )
  }
}

/**
 * Validates that there are no blank lines within a specific line range in strict mode.
 */
export function validateNoBlankLinesInRange(
  startLine: number,
  endLine: number,
  blankLines: BlankLineInfo[],
  strict: boolean,
  context: string,
): void {
  if (!strict)
    return

  // Find blank lines within the range
  const firstBlank = blankLines.find(
    blank => blank.lineNumber > startLine && blank.lineNumber < endLine,
  )

  if (firstBlank) {
    throw new ToonDecodeError(
      `Blank lines inside ${context} are not allowed in strict mode`,
      { line: firstBlank.lineNumber },
    )
  }
}

// #endregion

// #region Row classification helpers

/**
 * Checks if a line is a data row (vs a key-value pair) in a tabular array.
 */
function isDataRow(content: string, delimiter: Delimiter): boolean {
  const colonPos = content.indexOf(COLON)
  const delimiterPos = content.indexOf(delimiter)

  // No colon = definitely a data row
  if (colonPos === -1) {
    return true
  }

  // Has delimiter and it comes before colon = data row
  if (delimiterPos !== -1 && delimiterPos < colonPos) {
    return true
  }

  // Colon before delimiter or no delimiter = key-value pair
  return false
}

// #endregion
